// vue-collapsed/dist/index.mjs
import { defineComponent as $, toRef as H, computed as x, ref as S, shallowRef as E, onMounted as I, watch as _, createBlock as L, openBlock as M, resolveDynamicComponent as U, normalizeStyle as O, withCtx as k, renderSlot as q, normalizeProps as z, guardReactiveProps as B } from "vue";
const R = "--vc-auto-duration", P = `height var(${R}) cubic-bezier(0.33, 1, 0.68, 1)`, p = { padding: 0 }, V = 300, Y = {
  position: "absolute",
  width: "1px",
  height: "1px",
  padding: "0",
  margin: "-1px",
  overflow: "hidden",
  clip: "rect(0, 0, 0, 0)",
  whiteSpace: "nowrap",
  border: "0"
};
function C(a) {
  return a.value ? parseFloat(getComputedStyle(a.value).height) : 0;
}
function D(a) {
  if (!a.value) return {};
  const { transition: n } = getComputedStyle(a.value);
  return n === "all 0s ease 0s" || n === "all" ? { transition: P } : { transition: n };
}
function b(a) {
  if (!a.value) return !0;
  const { transition: n } = getComputedStyle(a.value);
  return typeof window.requestAnimationFrame > "u" || window.matchMedia("(prefers-reduced-motion: reduce)").matches || n.includes("none") || n.includes("height 0s");
}
function K(a = 0) {
  if (a === 0) return 0;
  const n = a / 36, o = Math.round((4 + 15 * n ** 0.25 + n / 5) * 10);
  return Number.isFinite(o) ? o : 0;
}
const J = /* @__PURE__ */ $({
  inheritAttrs: !0,
  __name: "Collapse",
  props: {
    when: { type: Boolean },
    baseHeight: { default: 0 },
    as: { default: "div" }
  },
  emits: ["collapse", "expand", "collapsed", "expanded"],
  setup(a, { emit: n }) {
    const o = a, s = n, i = H(o, "when"), l = H(o, "baseHeight"), v = x(() => ({ overflow: "hidden", height: `${l.value}px` })), f = x(() => ({
      ...p,
      ...l.value === 0 ? { display: "none" } : v.value
    })), e = S(null), h = S(i.value ? "expanded" : "collapsed"), c = (t) => h.value = t;
    function N() {
      return i.value ? p : l.value === 0 ? Y : f.value;
    }
    const g = E(N()), u = (t) => g.value = t, r = (t) => u({ ...g.value, ...t }), w = S(V), T = (t) => w.value = t, A = x(() => ({ [R]: `${w.value}ms` }));
    let d = NaN;
    function m() {
      u(p), c("expanded"), s("expanded");
    }
    function y() {
      u(f.value), c("collapsed"), s("collapsed");
    }
    I(() => {
      if (!e.value) return;
      const t = K(e.value.scrollHeight - l.value);
      t > 0 && T(t), !i.value && l.value === 0 && u(f.value);
    }), _(i, (t) => {
      if (e.value)
        if (d = NaN, t) {
          if (b(e)) return m();
          c("expanding"), s("expand"), u({
            ...p,
            ...v.value,
            ...A.value
          }), requestAnimationFrame(() => {
            if (e.value) {
              if (e.value.scrollHeight === 0) return m();
              d = e.value.scrollHeight, r({
                ...D(e),
                height: `${d}px`,
                willChange: "height"
              });
            }
          });
        } else {
          if (b(e) || (c("collapsing"), s("collapse"), r({
            ...A.value,
            height: `${e.value.scrollHeight}px`
          }), e.value.scrollHeight === 0)) return y();
          requestAnimationFrame(() => {
            e.value && r({
              ...v.value,
              ...D(e),
              willChange: "height"
            });
          });
        }
    }), _(l, (t) => {
      i.value || (t > 0 ? r({ display: void 0, height: `${t}px` }) : r({ display: "none" }));
    });
    function F(t) {
      e.value && t.target && t.target === e.value && t.propertyName === "height" && (i.value ? Math.abs(e.value.scrollHeight - C(e)) < 1 ? m() : d < e.value.scrollHeight && r({
        height: `${e.value.scrollHeight}px`
      }) : Math.abs(l.value - C(e)) < 1 && y());
    }
    return (t, j) => (M(), L(U(o.as), {
      ref_key: "collapseRef",
      ref: e,
      style: O(g.value),
      onTransitionend: F,
      "data-collapse": h.value
    }, {
      default: k(() => [
        q(t.$slots, "default", z(B({ state: h.value })))
      ]),
      _: 3
    }, 40, ["style", "data-collapse"]));
  }
});
export {
  J as Collapse
};
